### DISCUSSION TITLE
[Types] unified query types of nextServer.render & GetServerSideProps

### DISCUSSION THREAD

**Discussion by bobaaaaa** [vercel/next.js#26433] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

We use nextjs (9.4) with TypeScript and a custom server. The routes are very simple: one `Article` page accepting a single `id` for fetching data and renders. This works pretty well. Both query objects in `nextServer.render` and `GetServerSideProps` are typed as `interface ParsedUrlQuery { [key: string]: string | string[]; }`. This is totally okay for a default case. Want I miss is a generic type for both where I can explicit allow query params.

I describe it here in code:

```ts
interface ParsedUrlQuery { [key: string]: string | string[]; }

export type GetServerSidePropsContext<
  Q extends ParsedUrlQuery = ParsedUrlQuery
> = {
  req: IncomingMessage
  res: ServerResponse
  params?: Q
  // feature request 1:
  // explicit type for `query` like `params`
  query: ParsedUrlQuery
  preview?: boolean
  previewData?: any
}

export type GetServerSideProps<
  P extends { [key: string]: any } = { [key: string]: any },
  Q extends ParsedUrlQuery = ParsedUrlQuery
> = (context: GetServerSidePropsContext<Q>) => Promise<{ props: P }>
```

```ts
// /packages/next/next-server/server/next-server.ts
public async render(
    req: IncomingMessage,
    res: ServerResponse,
    pathname: string,
    // feature request 2:
    // explicit type for `query`
    query: ParsedUrlQuery = {},
    parsedUrl?: UrlWithParsedQuery
  ): Promise<void>
```

Here some simplified code from our codebase (incl. the "problems"):

```ts
// src/pages/article.tsx
export interface ArticleProps {
  content: Article;
}
export type ServerSideParams = {
  id: string;
};
export type ServerSideArticle = GetServerSideProps<ArticleProps, ServerSideParams>;
export const getServerSideProps: ServerSideNewsArticle = async context => {
  // problem 1:
  // need an explicit cast to `string`
  // problem 2:
  // everything is a valid query param. No type reference to `ServerSideQueryParams`.
  // => error prune in case of refactorings
  const articleId = context.query.id as string;
  const article: Article = await fetchArticle(articleId);
  return { props: { article } };
};
```

```ts
// /server.ts
// fyi: const app: NextInstance = Next({ dev, dir });
router.get(`/${articleId}`, async ctx => {
  // problem 1:
  // everything is a valid query param.
  await app.render(ctx.req, ctx.res, '/article', { id: ctx.params.id });
});
```

## Describe the solution you'd like

Based on the example something like this would be nice:

```ts
// src/pages/article.tsx
export interface ArticleProps {
  content: Article;
}

// I only need the `query` types not the `params`. Idk what fits best here.
export type ServerSideParams = {
  id: string;
};
export type ServerSideArticleRenderQuery  = {
  id: string;
};

export type ServerSideArticle = GetServerSideProps<ArticleProps, ServerSideParams, ServerSideArticleRenderQuery>;
export const getServerSideProps: ServerSideNewsArticle = async context => {
  const articleId = context.query.id; // ok
  const foo = context.query.foo; // not ok (type error)
  const article: Article = await fetchArticle(articleId);
  return { props: { article } };
};
```

```ts
// /server.ts
// fyi: const app: NextInstance = Next({ dev, dir });
import type { ServerSideArticleRenderQuery } from './pages/article'
router.get(`/${articleId}`, async ctx => {
  // ok
  await app.render<ServerSideArticleRenderQuery>(ctx.req, ctx.res, '/article', { id: ctx.params.id }); 
  // not ok (type error)
  await app.render<ServerSideArticleRenderQuery>(ctx.req, ctx.res, '/article', { foo: ctx.params.id });
});
```

**Comment by bennettdams** [comment]
The issue for this feature request was closed without comment, is there any way to get feedback why we wouldn't have that new generic?


Here's some more input for why this is helpful:

I understand that the `query` is dynamic by nature and every user can change the URL as he pleases - but most of the time, we as developers are creating the query for our `getServerSideProps` to consume. Having a type here would still be better than "nothing" in my opinion. 
You can also use a parsing library (like Zod) for the query - these libraries bring your type with it for the validated result.
I just thought it would be nice if we at least can provide a shortcut (via generic) without the need of another library.

The only way I was able to have some type-safety is by using `as` - which is always a bad idea. Have you come up with other ways to use `query` in a type-safe manner?


Here's an example. 
Imagine a route that will always need a query to work.
First, I check the raw query for existence and check whether it is an object, and after that overwrite the type with `as`:

```ts
type Query = ParsedUrlQuery & {
  someId: string;
  someNumber: number;
};

type Result = {
  someResult: string;
};

export const getServerSideProps: GetServerSideProps<Result> = async (
  context
) => {
  const queryRaw = context.query;

  // check raw query without knowing the type
  if (!!queryRaw && typeof queryRaw === "object") {
    // overwrite the type with `as` :(
    const query: Query = context.query as Query;

    const someId = query.someId;
    const someNumber = query.someNumber;

    return {
      props: {
        someResult: someId + someNumber,
      },
    };
  } else {
    return {
      notFound: true,
    };
  }
};
```

Are there better ways to use `query` in a type-safe manner without `as`?

As described above, the problem lies in the generics of the `GetServerSidePropsContext` of Next.js' types. There, `params`' type can be set via a generic type, which is passed by the `GetServerSideProps` type, but the `query` does not take a generic, it is just of type `ParsedUrlQuery`.

**Couldn't we just allow `GetServerSidePropsContext` to also take a generic for the `query`, like we do with `params`?** 

The only downside I can think of is that [a developer cannot skip generics when there are multiple available](https://github.com/microsoft/TypeScript/issues/10571) right now. Which means that there are situations where some developers have to give a generic that they don't have to use.

e.g. here, I added a new generic (`TQuery`) for the `query`

```ts
export type GetServerSideProps<
  // result tyope
  P extends { [key: string]: any } = { [key: string]: any },
  // params type
  Q extends ParsedUrlQuery = ParsedUrlQuery,
  // NEW: query type
  TQuery extends ParsedUrlQuery = ParsedUrlQuery,
  // previewData type
  D extends PreviewData = PreviewData
> = (
```

If you only want to use `params` and `previewData` in your code, you'd still need to define the type for `query`, even though you don't use it.

In my opinion, this trade off is worth the type-safety for the majority of developers, assuming that more developers use `query` than `previewData`, because you can just omit the generic for `previewData`, as it is in the last place of the generics.

Would this be worth a PR?
